nmdc_runtime/api/db/mongo.py

```diff
@@ -73,3 +78,107 @@ def activity_collection_names(mdb: MongoDatabase) -> Set[str]:
         "functional_annotation_set",
         "genome_feature_set",
     }
+
+
+def all_docs_have_unique_id(coll) -> bool:
+    first_doc = coll.find_one({}, ["id"])
+    if first_doc is None or "id" not in first_doc:
+        # short-circuit exit for empty collection or large collection via first-doc peek.
+        return False
+
+    total_count = coll.count_documents({})
+    return len(coll.distinct("id")) == total_count
+
+
+def ensure_unique_id_indexes(mdb: MongoDatabase):
+    """Ensure that any collections with an "id" field have an index on "id"."""
+    candidate_names = (
+        set(mdb.list_collection_names()) | schema_collection_names_with_id_field()
+    )
+    for collection_name in candidate_names:
+        if collection_name.startswith("system."):  # reserved by mongodb
+            continue
+
+        if (
+            collection_name in schema_collection_names_with_id_field()
+            or all_docs_have_unique_id(mdb[collection_name])
+        ):
+            mdb[collection_name].create_index("id", unique=True)
+
+
+class UpdateStatement(BaseModel):
+    q: dict
+    u: dict
+    upsert: bool = False
+    multi: bool = False
+
+
+class DeleteStatement(BaseModel):
+    q: dict
+    limit: conint(ge=0, le=1) = 1
+
+
+class OverlayDBError(Exception):
+    pass
+
+
+class OverlayDB(AbstractContextManager):
```

Comment:
Can you add a couple of comments explaining what the OverlayDB class is for

Reply:
Done.